### ISSUE TITLE
IS NOT NULL / IS DISTINCT FROM NULL should use index

### ISSUE THREAD

**Issue by EPluribusUnum** [FirebirdSQL/firebird#7797]
Hi *!

SELECT * FROM customer c WHERE c.country IS NOT NULL
SELECT * FROM customer c WHERE c.country IS DISTINCT FROM NULL

both PLAN is NATURAL.

for every data type there is an equivalent expression what uses index.
varchar : field >= ''
numeric : (field >=0 OR field < 0); or field>= "possible minimum value"
date/time : like numeric
boolean : field IN (TRUE, FALSE)

**Comment by EPluribusUnum** [comment]
Yes, in our case a lot of our fields (with foreign keys) contains mostly NULLs.
To utilize FK indexes corrently we write the equivalent expressions, but it is not as expressive than the IS NOT NULL/IS DISTINCT FROM condition.
Maybe then the next ODS version could keep statistics about null/not null ration as well.

**Comment by tomaszdubiel18** [comment]
Yeah, this topic has been already a few times mentioned in firebird support google group. In our case it's also quite frequent to have a lot of nulls for columns used in indexes. Firebird's ability to know about null selectivity would be very appreciated.

**Comment by mrotteveel** [comment]
Have you tried in Firebird 5.0 with a partial index?

**Comment by EPluribusUnum** [comment]
Partial index does not help. 
COMPUTED BY (country IS NOT NULL)
COMPUTED BY (country IS DISTINCT FROM NULL)

in both case the PLAN is still NATURAL

**Comment by dyemanov** [comment]
Partial index (in v5) is not he same as computed index.

`CREATE INDEX ... WHERE country IS NOT NULL`
`CREATE INDEX ... WHERE country IS DISTINCT FROM NULL`

**Comment by EPluribusUnum** [comment]
Oh, I see. Partial index helps, FB uses the new index.
But now the data is indexed twice, 1 index for FK and 1 partial index.

**Comment by mrotteveel** [comment]
That is a trade-off, yes, but unless and until we get better index statistics, this is probably the only workaround you can apply.

**Comment by dyemanov** [comment]
Not really twice, I'd say 1.x times, where x < 0.5 ;-) And this is just a workaround until NULLs are accounted in statistics.

**Comment by tomaszdubiel18** [comment]
After 2,5 years, are there some plans for this improvement? I think I described very similar problem here:
https://groups.google.com/g/firebird-support/c/UEg-Zaj3Stw
